035-cluster-wide-volume-usage-quota-managemnet.md

```diff
@@ -0,0 +1,280 @@
+are replicated to all brokers thus there is largely similar disk usage between all brokers. Assuming all the brokers
+have similar disk usage levels, they will all apply rate limits at similar times and levels, effectively giving cluster-wide out-of-storage protection.
+
+However, as clusters scale up the likelihood of even topic distribution drops. When topics are not evenly distributed it
+is possible for replication from broker `A` to broker `B` will cause broker `B` to consume all available disk space
+without triggering throttling of clients, as broker `A` disk usage remains acceptable.
+
+[KIP-73](https://cwiki.apache.org/confluence/display/KAFKA/KIP-73+Replication+Quotas) replication quotas are designed to
+manage the additional network load of migrating replicas between brokers, which does not address client publication
+leading to out of disk conditions through replication.
+
+Currently, the kafka-static-quota-plugin considers the total quantity of storage and how much of that is consumed (
+see [issue#2](https://github.com/strimzi/kafka-quotas-plugin/issues/2)) when considering whether to apply throttling to
+clients. This is problematic with respect
+to handling disk failure for JBODs [(KIP-112)](https://cwiki.apache.org/confluence/display/KAFKA/KIP-112%3A+Handle+disk+failure+for+JBOD) as the broker
+will take partitions offline when the volume they are stored on runs out of space. Which in the case of unbalanced usage
+between volumes can lead to a volume running out of storage without throttling being applied. The broker will go offline
+if all volumes are unavailable.
+
+Configuring the soft and hard limit thresholds is currently done in terms of bytes consumed, which forces users
+deploying the plugin to calculate a threshold for throttling and would require separate thresholds per volume to avoid
+out of space issues.
+
+## Proposal
+### Leverage Kafka to distribute volume usage metrics throughout the cluster.
+By publishing per volume usage metrics to a compacted topic keyed by the broker ID, each instance of the plugin will be able to start and quickly determine the state of the cluster and consistently apply throttling regardless of where the
+client is connected.
+
+[KIP-257](https://cwiki.apache.org/confluence/display/KAFKA/KIP-257+-+Configurable+Quota+Management) describes configurable quota management, where quotas are defined in terms of how much of a delay to apply to a given produce request. This leads to defining the
+possible states of the plugin as:
+`OPEN`, `THROTTLE` and `PAUSE` to accurately reflect the effects of transitioning between states.
+
+This implies there are three roles.
+1. **Data source** - Publish volume usage metrics about the volumes underlying the brokers log dirs.
+2. **Data Sink** - Consume volume usage metrics and make those available for making Quota Policy decisions.
+3. **Quota Policy** - Convert volume usage metrics for the entire cluster into throttling decisions.
+
```

Comment:
I think a [communication diagram](https://en.wikipedia.org/wiki/Communication_diagram) showing the interaction between components running over two brokers would be good.

Reply:
I've added a diagram to the proposal. Its not a diagram format I have any experience with so keen to hear any feedback on it.